Data models for Component-level Versioning Only

ormsbee opened the pull request:
# High Level Overview

This is a pared down version of the last experimental pull request, with Unit-level work and top level versioning of the entire LearningContext stripped out. A lot of other changes as well.

The `load_components` management command is functional and should pull in the XBlock contents for HTML, Problem, and Video types. However it doesn't yet pull in all the references files (e.g. HTML files, subtitle files, etc.) yet. It will pull in regex matched static asset references from HTML and Capa, but not things where it has to understand conventions based on XML attribute values (like how some video subtitles are specified).

## Apps

There are two apps at work here:

* `openedx_learning.core.publishing` has just enough to tie content together (LearningPackage) and to track when things get published (PublishLogEntry).
* `openedx_learning.core.components` has the models needed to track Components, their versioning, the raw data they address, and the publishing status of those models. It depends on the publishing app.

## Versioning

Versioning happens at the level of each Component (e.g. an individual block, like a problem or video) via the ComponentVersion model.

The Content model just holds raw bytes with some very minimal metadata (size, MIME type). Content has a M:M relationship with ComponentVersion via the custom ComponentVersionContent model (which provides an identifier that we use like a filename). That allows us to have the same Content associated with multiple ContentVersions if we want, making identifier renames cheap. The `load_components` management command will associate static assets with the appropriate ContentVersions.

The latest published version of a component is tracked in PublishedComponent, with a historical log available in ComponentPublishLogEntry which is M:1 with the `publish` app's PublishLogEntry. This allow us to model "a bunch of ComponentVersions were published at the same time".

Overall, the shift to versioning at a more granular level has made the modeling simpler, faster, and with less per-version overhead. It also makes diffs easier. That comes at a cost of making "point-in-time" snapshots more difficult to construct (you'd have to look at ComponentPublishLogEntry (joined with PublishLogEntry) to look for the most recent `component_version` before a certain time for each `component`. Certainly possible, but more expensive.

## Status

* there is no `data.py` defined
* there is no `api.py` layer at the moment, only direct model access.

ormsbee commented:
FYI @bradenmacdonald

ormsbee commented:
FYI @feanil, @kdmccormick

kdmccormick reviewed (approved):
@ormsbee @feanil You guys were right, reading the description made this all very clear.

I like that you've scoped this down to just Content and Component for now. I think that models here are extremely reasonable and will be a solid base as we work our way back up the hierarchy.

I had been resisting Component-level versioning for a while, but in retrospect trying to do Segment- or Unit-level versioning just made things more complex. I'm not sure why I felt so strongly--maybe because allowing library components to be versioned separately sounds like a UX nightmare? Even so, the Library Authoring system could mandate that library components are versioned together, without that restriction having to appear at this level.

Just some nits/thoughts below but nothing blocking.

ormsbee commented on openedx_learning/core/components/models.py:
```diff
@@ -0,0 +1,295 @@
+    """
+
+    uuid = immutable_uuid_field()
+    component = models.ForeignKey(Component, on_delete=models.CASCADE)
```
No, because many ComponentVersions might reference the same Component.

ormsbee commented on openedx_learning/core/components/models.py:
```diff
@@ -0,0 +1,295 @@
+    #    on_delete=models.SET_NULL,
+    #
+
+    contents = models.ManyToManyField(
+        "Content",
+        through="ComponentVersionContent",
+        related_name="component_versions",
+    )
```
Huh... I guess it's possible that we'd make one without Content (like just a title). I'd rather make the admin read-only altogether though, because ComponentVersion creation should include adding Publish entries rather than just direct model manipulation (I guess I really should do the api.py pieces of this soon).

ormsbee commented on openedx_learning/core/components/models.py:
```diff
@@ -0,0 +1,295 @@
+    component = models.OneToOneField(
+        Component, on_delete=models.RESTRICT, primary_key=True
+    )
+    component_version = models.ForeignKey(
```
I think I did it this way because I didn't realize you could have a nullable OneToOneField. I'll change it.

ormsbee commented on openedx_learning/core/components/models.py:
```diff
@@ -0,0 +1,295 @@
+    hash_digest = hash_field()
+
+    # Per RFC 4288, MIME type and sub-type may each be 127 chars.
+    type = models.CharField(max_length=127, blank=False, null=False)
+    sub_type = models.CharField(max_length=127, blank=False, null=False)
```
How would you feel about `media_type` and `media_subtype`?

kdmccormick replied:
Great 👍🏻

bradenmacdonald replied:
What's the rationale for splitting media type into two separate fields? Seems a bit strange to me.

kdmccormick replied:
We had some brief discussion here: https://github.com/openedx/openedx-learning/pull/13#discussion_r937034472. 

I thought splitting them felt right, but I don't feel strongly about it, especially since I'm not sure what the mime types will be used for. Like, I imagine that if we want to be able to filter the table for all images or all text, it's more efficient to do that if the top-level `media_type` is its own column, right?

On the other hand, splitting the mime type into two columns does seem like the _less_-obvious schema, so if we can't think of a compelling reason to do that, then keeping them as a single column sounds good to me.

feanil replied:
I think being able to filter on the `media_type` seems like a useful feature but are there use-cases of it that we would need for content creation or presentation?  Or is it more useful for analytics? If it's just analytics then I might suggest that we leave it as one top-level `mime-type` here because any analytics copy of this data can probably add the extra columns it needs to make analysis easier.

ormsbee replied:
I'm inclined to keep them separate, since it will make certain queries easier–like facet count by media_type for all images, as @kdmccormick points out, and we can always join them together for presentation purposes if we want.

Though this reminds me that I need to put proper indexing on them. :-/ I'll do that now...

bradenmacdonald replied:
My assumption was that if you aren't indexing the fields anyways then it doesn't much matter if they're split or not - the performance will be similar because the DB still has to scan over all the values. On the other hand, if you use a single combined field but you _are_ indexing it, then prefix filtering like counting all the values that start with `image/` is essentially just as fast as if it's a unique value in its own column, because prefix matching with a text column index is extremely fast. (The only case where I see the split columns being useful would actually be if you needed to filter on the _second_ part of the media type without regard to the first part / prefix... but that doesn't seem to be a use case).

I also think that the prefixes aren't that meaningful in a lot of cases... Some might consider JSON to be a text value but it's in the `application/` prefix, and OLX is in neither `text/` nor `application/` as you might expect but `vnd/`. CSS and HTML, unlike JSON, are in the `text/` prefix; except XHTML (now deprecated) which is in `application/`. In short, the only prefix that I think is remotely useful or consistent is `image/`, and you can very easily and efficiently filter and count based on that prefix if there's an index on a combined text `media_type` column.

Anyhow I think either way is fine with me; I don't think it's a big deal at all, but I am also not convinced there's any benefit to splitting them up, while it is slightly more complex.

kdmccormick replied:
> prefix matching with a text column index is extremely fast

Good to know. With that in mind, I'm fully in favor of a single field.

kdmccormick replied:
Sorry, let me be more clear: any reservations I had about a single field are resolved. Two fields is also still OK with me. Agreed that this probably a very minor decision.

ormsbee replied:
Okay, I made it one field.

bradenmacdonald commented:
Hah, I was just thinking about this recently and I ended up concluding to myself that component-level versioning seemed like the way to go. Glad to see you trying it out and to hear it seems to be proving out so far from what you've tried.

bradenmacdonald commented on docs/decisions/0002-content-flexibility.rst:
```diff
@@ -18,18 +18,11 @@ Decision
 
 Unit
-  This is a list of one or more Segments that is displayed to the user on one page. A Unit may be stitched together using content that comes from multiple sources, such as content libraries. Units do not have to be strictly instructional content, as things like upgrade offers and error messages may also be injected.
+  A Unit is an ordered list of one or more Components. A Unit is addressable in the browser at some URL, and is displayed together. A common use case might be to display some introductory Text, a Video, and some followup Problem (all separate Components). An individual Component in a Unit may or may not make sense when taken outside of that Unit–e.g. a Video may be reusable elsewhere, but the Pro ...
```
> A Unit is addressable in the browser at some URL

Does this imply that a Component is _not_ accessible at some URL in general?

ormsbee replied:
I think this was a leftover from when Segments were not going to necessarily be separately addressable. 

I'm honestly not sure how I feel about Components being separately rendered vs. having implicit Units with single Components in them as the thing being rendered. I like the idea of having one type of thing being rendered at that level, and the idea that some Components aren't meant to be rendered in isolation. But that's super-hand-wavy at this point, so I'll just remove this wording and kick that decision down the road a little.

ormsbee commented:
@kdmccormick, @feanil, @bradenmacdonald: I think I addressed all the comments. Also added a couple small conveniences to the Django admin, a sprinkling of indexes, and some more fleshed out comments.

bradenmacdonald commented on openedx_learning/core/components/models.py:
```diff
@@ -0,0 +1,295 @@
+    # is "xblock.v1" (to match the setup.py entrypoint naming scheme).
+    namespace = models.CharField(max_length=100, null=False, blank=False)
+
+    # type is a way to help sub-divide namespace if that's convenient. This
+    # field cannot be null, but it can be blank if it's not necessary. For an
+    # XBlock, type corresponds to tag, e.g. "video". It's also the block_type in
+    # the UsageKey.
+    type = models.CharField(max_length=100, null=False, blank=True)
```
If we expect that the only actual usage of this field will be for subdividing the ID namespace for XBlocks to match the historical behavior, then we could conceivably just call this field `block_type`. I don't think it _should_ be used outside of XBlocks, as having two levels of namespacing is not ideal.

bradenmacdonald commented on openedx_learning/core/components/models.py:
```diff
@@ -0,0 +1,403 @@
+    about scalability issues. For instance, video files should not be stored
+    here directly. There is a 10 MB limit set for the moment, to accomodate
+    things like PDF files and images, but the itention is for the vast majority
+    of rows to be much smaller than that.
```
How _should_ data files over 10 MB be stored?

I would almost prefer a much lower limit like 250 kB and an integrated way to store and reference larger Content instances as objects on S3. I suspect you may already have implemented or planned something that but I forgot or don't see it yet.

ormsbee replied:
You're right, I've punted on this issue. I came to the same conclusion that you did: S3 objects (probably being django-storages) + database references, probably through an optional FileField on Content.

Captured this in https://github.com/openedx/openedx-learning/issues/29

bradenmacdonald commented on openedx_learning/core/components/readme.rst:
```diff
@@ -0,0 +1,22 @@
+Architecture Guidelines
+-----------------------
+
+* We're keeping nearly unlimited history, so per-version metadata (i.e. the space/time cost of making a new version) must be kept low.
```
I don't suppose there's any way to tell which versions are in use or not with this model, because any application may or may not have foreign keys to specific `ComponentVersion`s?

(I'm wondering if we can auto-prune unused versions.)

ormsbee replied:
I think we can using Django ORM introspection or a mixin, but it wouldn't be perfect. I'm honestly hoping the per-version overhead is low enough now where it's okay not to prune, because that just makes content lifecycle management so much easier.

bradenmacdonald reviewed (approved):
@ormsbee This is excellent and I really like the direction. I don't have any major feedback, just a couple little questions. Nothing blocking. Looking forward to building on this!
